tools/clang/unittests/HLSLExec/LongVectors.h

```diff
@@ -199,68 +73,375 @@ void FillLongVectorDataFromShaderBuffer(MappedData &ShaderBuffer,
   }
 }
 
-enum LongVectorOpType {
-  LongVectorOpType_ScalarAdd,
-  LongVectorOpType_ScalarMultiply,
-  LongVectorOpType_Multiply,
-  LongVectorOpType_Add,
-  LongVectorOpType_Min,
-  LongVectorOpType_Max,
-  LongVectorOpType_Clamp,
-  LongVectorOpType_Initialize,
-  LongVectorOpType_UnInitialized
+template <typename T> constexpr bool IsFloatingPointType() {
+  return std::is_same_v<T, float> || std::is_same_v<T, double> ||
+         std::is_same_v<T, HLSLHalf_t>;
+}
+
+struct LongVectorOpTypeStringToEnumValue {
+  std::wstring OpTypeString;
+  uint32_t OpTypeValue;
+};
+
+template <typename T>
+T GetLongVectorOpType(const LongVectorOpTypeStringToEnumValue *Values,
+                      const std::wstring &OpTypeString, std::size_t Length) {
+  for (size_t i = 0; i < Length; i++) {
+    if (Values[i].OpTypeString == OpTypeString) {
+      return static_cast<T>(Values[i].OpTypeValue);
+    }
+  }
+
+  LOG_ERROR_FMT_THROW(L"Invalid LongVectorOpType string: %s",
+                      OpTypeString.c_str());
+
+  return static_cast<T>(UINT_MAX);
+}
+
+namespace LongVector {
+
+enum ValidationType {
+  ValidationType_Epsilon,
+  ValidationType_Ulp,
+};
+
+enum BasicOpType {
+  BasicOpType_Binary,
+  BasicOpType_Unary,
+  BasicOpType_ScalarBinary,
+  BasicOpType_EnumValueCount
+};
+
+enum BinaryOpType {
+  // Below this line to be moved to HLSLOpType (+, -, * etc)
+  BinaryOpType_ScalarAdd,
+  BinaryOpType_ScalarMultiply,
+  BinaryOpType_ScalarSubtract,
+  BinaryOpType_ScalarDivide,
+  BinaryOpType_ScalarModulus,
+  BinaryOpType_Multiply,
+  BinaryOpType_Add,
+  BinaryOpType_Subtract,
+  BinaryOpType_Divide,
+  BinaryOpType_Modulus,
+  // Below this line to be moved to HLSLMathOpType (min, max etc)
+  BinaryOpType_Min,
+  BinaryOpType_Max,
+  BinaryOpType_ScalarMin,
+  BinaryOpType_ScalarMax,
+  BinaryOpType_EnumValueCount
+};
+
+static const LongVectorOpTypeStringToEnumValue BinaryOpTypeStringToEnumMap[] = {
```

Comment:
Couldn't these be a `std::map`?

Reply:
I wanted to have a static assert on the size to have a compile error if you forget to add to this enum value when adding an *OpType. Even declaring the std::map as static const still wouldn't allow a static assert on the size.

Open to ideas if you think some other pattern is cleaner though.